[opaque obj] Error for torch.library.custom_op infer_schema

angelayi opened the pull request:
Unsure how we can get infer_schema to infer the scriptObject type from just the type annotation, so for now will just error clearly and ask users to specify a schema.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #163936
* #163714
* #163284
* #163278
* __->__ #163277
* #163279

angelayi committed:
[opaque obj] Error for torch.library.custom_op infer_schema

[ghstack-poisoned]

angelayi committed:
Update on "[opaque obj] Error for torch.library.custom_op infer_schema"

Unsure how we can get infer_schema to infer the scriptObject type from just the type annotation, so for now will just error clearly and ask users to specify a schema.




[ghstack-poisoned]

zou3519 commented on torch/_library/infer_schema.py:
```diff
@@ -124,7 +124,13 @@ def unstringify_type(ty: Union[type[object], str]) -> tuple[typing.Any, bool]:
 
         if annotation_type not in SUPPORTED_PARAM_TYPES:
-            if annotation_type.__origin__ is tuple:
+            if annotation_type == torch._C.ScriptObject:
```
I think we should be able to get this to work.
The idea is to provide some new NewType annotation that behaves like "ScriptObject" (or Any) during typing, but can be uniquely identified by this function.

https://mypy.readthedocs.io/en/stable/more_types.html#newtypes

angelayi committed:
Update on "[opaque obj] Error for torch.library.custom_op infer_schema"

Unsure how we can get infer_schema to infer the scriptObject type from just the type annotation, so for now will just error clearly and ask users to specify a schema.




[ghstack-poisoned]

angelayi committed:
Update on "[opaque obj] Error for torch.library.custom_op infer_schema"

Unsure how we can get infer_schema to infer the scriptObject type from just the type annotation, so for now will just error clearly and ask users to specify a schema.




[ghstack-poisoned]

angelayi committed:
Update on "[opaque obj] Error for torch.library.custom_op infer_schema"

Unsure how we can get infer_schema to infer the scriptObject type from just the type annotation, so for now will just error clearly and ask users to specify a schema.




[ghstack-poisoned]

angelayi committed:
Update on "[opaque obj] Error for torch.library.custom_op infer_schema"

Unsure how we can get infer_schema to infer the scriptObject type from just the type annotation, so for now will just error clearly and ask users to specify a schema.




[ghstack-poisoned]

angelayi committed:
Update on "[opaque obj] Error for torch.library.custom_op infer_schema"

Unsure how we can get infer_schema to infer the scriptObject type from just the type annotation, so for now will just error clearly and ask users to specify a schema.




[ghstack-poisoned]

angelayi commented on torch/library.py:
```diff
@@ -41,6 +42,7 @@
     "triton_op",
     "wrap_triton",
     "infer_schema",
+    "OpaqueType",
```
yes theyre not private yet!

angelayi committed:
Update on "[opaque obj] Error for torch.library.custom_op infer_schema"

Unsure how we can get infer_schema to infer the scriptObject type from just the type annotation, so for now will just error clearly and ask users to specify a schema.




[ghstack-poisoned]
